### PR TITLE
Kafka ProducerInterceptor Bean

### PR DESCRIPTION
**Pull request by ozangunalp**
Kafka connector outgoing channel attribute `interceptor-bean` for configuring a managed interceptor bean.

### CODE REVIEW

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java [hunk 1]
```diff
@@ -147,18 +159,32 @@ public Uni<Void> runOnSendingThread(java.util.function.Consumer<Producer<K, V>>
...
+                        intercepted = interceptor.onSend(record);
                     } else {
-                        em.complete(metadata);
+                        intercepted = record;
                     }
-                })));
+                    return Uni.createFrom().emitter(em -> c.send(intercepted, (metadata, exception) -> {
+                        if (exception != null) {
+                            if (interceptor != null) {
+                                interceptor.onAcknowledgement(getRecordMetadataForFailure(intercepted), exception);
```
Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java:172]: Wondering if we need a onNegativeAcknowledgement method instead (to be clear)

Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java:172]: What if the method throws an exception? Probably better to handle that case.

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java [hunk 2]
```diff
@@ -147,18 +159,32 @@ public Uni<Void> runOnSendingThread(java.util.function.Consumer<Producer<K, V>>
...
+                            }
+                            if (record.topic() != null) {
+                                log.unableToWrite(this.channel, record.topic(), exception);
+                            } else {
+                                log.unableToWrite(this.channel, exception);
+                            }
+                            em.fail(exception);
+                        } else {
+                            if (interceptor != null) {
+                                interceptor.onAcknowledgement(metadata, null);
```
Comment (cescoffier) [inline smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java:182]: What if the interceptor throws an exception?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java (hunk 1 above)
Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java:172]: The idea is to be able to use the standard ProducerInterceptor interface from the Kafka package.

Comment (ozangunalp) [reply smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/ReactiveKafkaProducer.java:172]: The interceptor is called inside the uni emitter. So a failure will be emitted, it'll be considered like a send error retried etc. finally nacking the message

**Comment by ozangunalp** [comment]
@cescoffier I decided to log and ignore the exceptions as documented in `ProducerInterceptor` `onSend` and `onAcknowledge` methods.

**Comment by ozangunalp** [comment]
Since the update revapi filters behave oddly.

**Review by cescoffier — APPROVED** [review]

**Pull request merged**